### PR TITLE
dialects: (onnx) add onnx.Constant

### LINKED ISSUE #1919: add onnx.Constant op

### PR DESCRIPTION
**Pull request by kayode-gif**
resolves #1919

### CODE REVIEW

File: tests/filecheck/dialects/onnx/onnx_ops.mlir [hunk 1]
```diff
@@ -60,3 +60,6 @@
 
 %res_conv_6 = "onnx.Conv"(%t23, %t24, %t25) {onnx_node_name = "/Conv", "auto_pad" = "NOTSET", "group" = 1 : i64, "kernel_shape" = [3 : i64, 3 : i64], "dilations" = [1 : i64, 1 : i64], "strides" = [2 : i64, 2 : i64], "pads" = [1 : i64, 0 : i64, 1 : i64, 0 : i64]}: (tensor<1x1x7x5xf32>, tensor<1x1x3x3xf32>, none) -> tensor<1x1x4x2xf32>
 //CHECK: %res_conv_6 = onnx.Conv(%t23, %t24, %t25) {"onnx_node_name" = "/Conv", "auto_pad" = "NOTSET", "group" = 1 : i64, "kernel_shape" = [3 : i64, 3 : i64], "dilations" = [1 : i64, 1 : i64], "strides" = [2 : i64, 2 : i64], "pads" = [1 : i64, 0 : i64, 1 : i64, 0 : i64]}: (tensor<1x1x7x5xf32>, tensor<1x1x3x3xf32>, none) -> tensor<1x1x4x2xf32>
+
+%res_constant = "onnx.Constant"() {onnx_node_name = "/Constant", "value" = dense<1> : tensor<1xi64>}: () -> tensor<1xi64>
+//CHECK: %res_constant = onnx.Constant() {"onnx_node_name" = "/Constant", "value" = dense<1> : tensor<1xi64>}: () -> tensor<1xi64>
```
Comment (superlopuh) [inline tests/filecheck/dialects/onnx/onnx_ops.mlir:65]:
```suggestion
%res_constant = "onnx.Constant"() {onnx_node_name = "/Constant", "value" = dense<1> : tensor<1xi64>}: () -> tensor<1xi64>
//CHECK: %res_constant = onnx.Constant() {"onnx_node_name" = "/Constant", "value" = dense<1> : tensor<1xi64>}: () -> tensor<1xi64>

```

File: xdsl/dialects/onnx.py [hunk 2]
```diff
@@ -570,11 +574,99 @@ def verify_(self) -> None:
...
+                "value_strings": value_strings,
+            },
+            operands=[],
+            result_types=[output_type],
+        )
+
+    def verify_(self) -> None:
+        if not isinstance(output_type := self.output.type, TensorType):
+            assert False, "onnx.Constant result must be of type TensorType"
+
```
Comment (superlopuh) [inline xdsl/dialects/onnx.py:644]: This check is not necessary, as it's already added automatically with the output result_def

**Comment by superlopuh** [comment]
Can you please fill in the fields in the side bar, with the reviewers, etc, when this PR is ready for review?

**Comment by kayode-gif** [comment]
> Can you please fill in the fields in the side bar, with the reviewers, etc, when this PR is ready for review?

okay will add additional reviewers now

**Comment by superlopuh** [comment]
Can you please also add the appropriate tag and project?

**Comment by kayode-gif** [comment]
> Can you please also add the appropriate tag and project?
sorry, it says "no projects" when i try add something

**Comment by superlopuh** [comment]
Ok I'll do that bit from now on, thanks for letting me know

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/onnx.py [hunk 3]
```diff
@@ -570,11 +574,94 @@ def verify_(self) -> None:
...
+                "value_int element type has to be a 64-bit signless integer"
+            )
+
+        if self.value_ints is not None:
+            for value in self.value_ints:
+                width = value.type.width.data
+                if width != 64:
+                    raise VerifyException(
+                        "value_ints elements type has to be a 64-bit signless integer"
+                    )
```
Comment (superlopuh) [inline xdsl/dialects/onnx.py:656]: We should probably check that exactly one value is provided also

Comment (kayode-gif) [reply xdsl/dialects/onnx.py:656]: done

File: xdsl/dialects/onnx.py [hunk 4]
```diff
@@ -570,11 +574,109 @@ def verify_(self) -> None:
...
+            self.value_floats,
+            self.value_int,
+            self.value_ints,
+            self.value_string,
+            self.value_strings,
+        ]
+        used_attrs = sum(1 for attr in attrs if attr is not None)
+        if used_attrs != 1:
+            raise VerifyException(
+                "More than one value atrribute provided. Only one should be used"
```
Comment (compor) [inline xdsl/dialects/onnx.py:670]:
```suggestion
               f"Only one value attribute must be provided, but {used_attrs} were specified"
```

if you have the information, make it explicit; moreover, the previous message was repeating itself a bit

**Review by compor — APPROVED** [review]

File: xdsl/dialects/onnx.py (hunk 4 above)
Comment (kayode-gif) [reply xdsl/dialects/onnx.py]: okay, thanks for the spot :)

**Pull request merged**